docs/architecture/adr-022-abci-errors.md

```diff
@@ -0,0 +1,60 @@
+# ADR 023: ABCI Codespaces
+
+## Changelog
+
+- *2018-09-01* Initial version
+
+## Context
+
+ABCI errors should provide an abstraction between application details
+and the client interface responsible for formatting & displaying errors to the user.
+
+Currently, this abstraction consists of a single integer (the `code`), where any
+`code > 0` is considered an error (ie. invalid transaction) and all type
+information about the error is contained in the code. This integer is
+expected to be decoded by the client into a known error string, where any
+more specific data is contained in the `data`.
+
+In a [previous conversation](https://github.com/tendermint/abci/issues/165#issuecomment-353704015),
+it was suggested that not all non-zero codes need to be errors, hence why it's called `code` and not `error code`.
+It is unclear exactly how the semantics of the `code` field will evolve, though
+better lite-client proofs (like discussed for tags
+[here](https://github.com/tendermint/tendermint/issues/1007#issuecomment-413917763))
+may play a role.
+
+For now, note that having all type information in a single integer
+places an unfortunate requirement on all users of the `code` system to
+co-ordinate up-front on which `code` means what, hence restricting their ability experiment
+with codes.
+
+For instance, while trying to build an ecosystem of modules that can be composed into a single
+ABCI application, the Cosmos-SDK had to hack a higher level "codespace" into the
+single integer so that each module could have its own space to express its
+errors.
+
+## Decision
+
+Include a `string code_space` in all ABCI messages that have a `code`.
+This allows applications to namespace the codes so they can experiment with
+their own code schemes.
+
+It is the responsibility of applications to limit the size of the `code_space`
```

Comment:
Why not use amino's method for going from code_space -> byte prefix. (While maybe removing the null byte restriction)

Reply:
Not following - can you expand ?

Reply:
amino takes the hash of the prefix you provide it, truncates it and removes null bytes, in order to get the prefix it uses for the registered concrete struct. I think we could do something similar for going from code_space string to prefix, just to remain more consistent. (Though this may just be even more confusing, I'm unsure)

Reply:
not sure what prefix you're talking about here for the code space, maybe that's why I'm confused

Reply:
>  Include a `string code_space` in all ABCI messages that have a `code`. 

I'm talking about that `code_space`. My understanding is were deriving domain seperation from this code_space string.

Reply:
right - what does it have to do with amino ?

Reply:
I was just suggesting to use the same algorithm to go from string -> prefix for domain separation.

Reply:
Ok maybe I'm still missing something but seems confusing to make this field contain opaque bytes based on amino

Reply:
Your right, that probably would be even more confusing.